### PR TITLE
Fix Ambigious Logs Format and Add Support for Details in Service Logs

### PR DESCRIPTION
**Pull request by dperny**
**- What I did**

### Please review carefully, the first commit in this series contains a change to the long-standing docker logs API.

3 commits, separate for ease of comparison:

1. **Update the logs API format to URL-encode characters in details** in docker logs. This allows the output of `docker logs` and `docker service logs` to be reliably parsed by machines, even if users put things like spaces and commas in details. See #31956 for a more technical explanation.

2. Update swarmkit vendoring to docker/swarmkit@1be67ab7, which contains a change to protobufs allowing the sending of details

3. Update the engine and CLI to support details in service logs. Add a `--pretty` flag to service logs, which is enabled by default, and which controls pretty formatting for service logs. `--pretty` flag is incompatible with TTY logs, because unusual TTY logs cannot be reliably parsed.

I will write docs after design review, to avoid doing a bunch of docs work that might get tossed if this design is bad. 

**- How I did it**

The only big thing to explain is `--pretty` and `--details`. 

Docker logs supports something called `details` (or, internally, `Attributes`), which is extra information affixed to the beginning of each log line (after the timestamp, if also requested). Details are a series of key-value pairs encoded in the format `key=value` and separated by commas. The details field is separated from the log line by a single space.

The `context` of service logs provides the node, service, and task IDs that the log line originates from. Under the hood, the context is sent to the user as part of the details. If the details are not requested, no context is sent. If the details *are* requested, context AND any details the user has provided will be sent. They are the same thing as far as the docker engine is concerned; only swarmkit makes a distinction. 

The `--pretty` flag is the default behavior of `docker service logs` before this change. It parses the context out of the details and presents it in a nice format to the user. If `--pretty=true`, which is the default, then the CLI will request the details as part of the log request. However, any extra details will not be displayed to the user, unless `--details` is ALSO passed.

If `--pretty=false`, then log lines will be shown but no context information will be included, unless `--details=true`, in which case, the details (including the context) will be printed unparsed.

TTY logs cannot be reliably parsed because of the mess of control characters that comes out of them, and so are incompatible with `--pretty=true`. To get logs from a service with a TTY, the user must pass `--pretty=false`.

Examples:

```
root@cdf7669f545a:/go/src/github.com/docker/docker# docker service create --name test -d=false --log-driver json-file --log-opt env=yourdetail -e yourdetail=yourvalue --replicas 3 flyinprogrammer/logtester
xubbghfyz6t0rz9yfz7ay8dxo

root@cdf7669f545a:/go/src/github.com/docker/docker# docker service logs --tail 5 test
test.2.sel8z2ew0onp@cdf7669f545a    | 2017-05-03 18:01:43.498367955 +0000 UTC stderr msg: 21
test.3.kzzla7vkybsf@cdf7669f545a    | 2017-05-03 18:01:43.006654046 +0000 UTC stderr msg: 21
test.3.kzzla7vkybsf@cdf7669f545a    | 2017-05-03 18:01:48.006965569 +0000 UTC stdout msg: 22
test.1.yekwyaw930u5@cdf7669f545a    | 2017-05-03 18:01:43.472566256 +0000 UTC stderr msg: 21
test.2.sel8z2ew0onp@cdf7669f545a    | 2017-05-03 18:01:48.499076705 +0000 UTC stdout msg: 22
test.2.sel8z2ew0onp@cdf7669f545a    | 2017-05-03 18:01:48.499076705 +0000 UTC stderr msg: 22
test.3.kzzla7vkybsf@cdf7669f545a    | 2017-05-03 18:01:48.006965569 +0000 UTC stderr msg: 22
test.1.yekwyaw930u5@cdf7669f545a    | 2017-05-03 18:01:48.473618498 +0000 UTC stderr msg: 22
test.1.yekwyaw930u5@cdf7669f545a    | 2017-05-03 18:01:48.473618498 +0000 UTC stdout msg: 22
test.2.sel8z2ew0onp@cdf7669f545a    | 2017-05-03 18:01:53.500325013 +0000 UTC stdout msg: 23
test.3.kzzla7vkybsf@cdf7669f545a    | 2017-05-03 18:01:53.007495332 +0000 UTC stdout msg: 23
test.3.kzzla7vkybsf@cdf7669f545a    | 2017-05-03 18:01:53.007495332 +0000 UTC stderr msg: 23
test.1.yekwyaw930u5@cdf7669f545a    | 2017-05-03 18:01:53.474327048 +0000 UTC stderr msg: 23
test.2.sel8z2ew0onp@cdf7669f545a    | 2017-05-03 18:01:53.500325013 +0000 UTC stderr msg: 23
test.1.yekwyaw930u5@cdf7669f545a    | 2017-05-03 18:01:53.474327048 +0000 UTC stdout msg: 23

root@cdf7669f545a:/go/src/github.com/docker/docker# docker service logs --tail 5 --pretty=false test
2017-05-03 18:02:18.48362645 +0000 UTC stdout msg: 28
2017-05-03 18:02:18.014483185 +0000 UTC stdout msg: 28
2017-05-03 18:02:18.507540495 +0000 UTC stderr msg: 28
2017-05-03 18:02:23.489279517 +0000 UTC stdout msg: 29
2017-05-03 18:02:23.016239588 +0000 UTC stderr msg: 29
2017-05-03 18:02:23.508265526 +0000 UTC stderr msg: 29
2017-05-03 18:02:23.489279517 +0000 UTC stderr msg: 29
2017-05-03 18:02:23.016239588 +0000 UTC stdout msg: 29
2017-05-03 18:02:23.508265526 +0000 UTC stdout msg: 29
2017-05-03 18:02:28.016786331 +0000 UTC stderr msg: 30
2017-05-03 18:02:28.493187361 +0000 UTC stderr msg: 30
2017-05-03 18:02:28.50953441 +0000 UTC stdout msg: 30
2017-05-03 18:02:28.493187361 +0000 UTC stdout msg: 30
2017-05-03 18:02:28.016786331 +0000 UTC stdout msg: 30
2017-05-03 18:02:28.50953441 +0000 UTC stderr msg: 30

root@cdf7669f545a:/go/src/github.com/docker/docker# docker service logs --tail 5 --details test
test.2.sel8z2ew0onp@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:48.511388777 +0000 UTC stdout msg: 34
test.1.yekwyaw930u5@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:48.502437728 +0000 UTC stderr msg: 34
test.3.kzzla7vkybsf@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:48.022797453 +0000 UTC stdout msg: 34
test.3.kzzla7vkybsf@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:53.023200542 +0000 UTC stderr msg: 35
test.2.sel8z2ew0onp@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:53.511614878 +0000 UTC stderr msg: 35
test.1.yekwyaw930u5@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:53.503153252 +0000 UTC stderr msg: 35
test.1.yekwyaw930u5@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:53.503153252 +0000 UTC stdout msg: 35
test.3.kzzla7vkybsf@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:53.023200542 +0000 UTC stdout msg: 35
test.2.sel8z2ew0onp@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:53.511614878 +0000 UTC stdout msg: 35
test.2.sel8z2ew0onp@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:58.511755244 +0000 UTC stdout msg: 36
test.1.yekwyaw930u5@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:58.508514003 +0000 UTC stderr msg: 36
test.3.kzzla7vkybsf@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:58.027852636 +0000 UTC stderr msg: 36
test.3.kzzla7vkybsf@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:58.027852636 +0000 UTC stdout msg: 36
test.2.sel8z2ew0onp@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:58.511755244 +0000 UTC stderr msg: 36
test.1.yekwyaw930u5@cdf7669f545a    | yourdetail=yourvalue 2017-05-03 18:02:58.508514003 +0000 UTC stdout msg: 36

root@cdf7669f545a:/go/src/github.com/docker/docker# docker service logs --tail 5 --pretty=false --details test
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=yekwyaw930u5u6vdfuzrvua18,yourdetail=yourvalue 2017-05-03 18:03:23.522596702 +0000 UTC stderr msg: 41
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=sel8z2ew0onpjwxvtrrvyntxp,yourdetail=yourvalue 2017-05-03 18:03:23.522580521 +0000 UTC stderr msg: 41
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=yekwyaw930u5u6vdfuzrvua18,yourdetail=yourvalue 2017-05-03 18:03:28.524276735 +0000 UTC stdout msg: 42
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=kzzla7vkybsf0dxhat8izesdl,yourdetail=yourvalue 2017-05-03 18:03:23.03757233 +0000 UTC stdout msg: 41
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=sel8z2ew0onpjwxvtrrvyntxp,yourdetail=yourvalue 2017-05-03 18:03:28.524254861 +0000 UTC stderr msg: 42
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=yekwyaw930u5u6vdfuzrvua18,yourdetail=yourvalue 2017-05-03 18:03:28.524276735 +0000 UTC stderr msg: 42
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=kzzla7vkybsf0dxhat8izesdl,yourdetail=yourvalue 2017-05-03 18:03:28.037789504 +0000 UTC stderr msg: 42
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=yekwyaw930u5u6vdfuzrvua18,yourdetail=yourvalue 2017-05-03 18:03:33.525308163 +0000 UTC stdout msg: 43
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=sel8z2ew0onpjwxvtrrvyntxp,yourdetail=yourvalue 2017-05-03 18:03:28.524254861 +0000 UTC stdout msg: 42
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=sel8z2ew0onpjwxvtrrvyntxp,yourdetail=yourvalue 2017-05-03 18:03:33.524724657 +0000 UTC stderr msg: 43
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=yekwyaw930u5u6vdfuzrvua18,yourdetail=yourvalue 2017-05-03 18:03:33.525308163 +0000 UTC stderr msg: 43
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=kzzla7vkybsf0dxhat8izesdl,yourdetail=yourvalue 2017-05-03 18:03:28.037789504 +0000 UTC stdout msg: 42
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=kzzla7vkybsf0dxhat8izesdl,yourdetail=yourvalue 2017-05-03 18:03:33.038901014 +0000 UTC stderr msg: 43
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=sel8z2ew0onpjwxvtrrvyntxp,yourdetail=yourvalue 2017-05-03 18:03:33.524724657 +0000 UTC stdout msg: 43
com.docker.swarm.node.id=auh7b8e919ylsv08bwnus25di,com.docker.swarm.service.id=xubbghfyz6t0rz9yfz7ay8dxo,com.docker.swarm.task.id=kzzla7vkybsf0dxhat8izesdl,yourdetail=yourvalue 2017-05-03 18:03:33.038901014 +0000 UTC stdout msg: 43
```

**- How to verify it**

A new integration test testing details has been written. Otherwise, give it a shot. The default behavior and CLI is unchanged from the previous version, if no flags are passed, except in the case of TTY logs, in which case `--pretty=false` must be passed.

**- Description for the changelog**

- Fix ambiguous logs format #31956
- Support log details in `docker service logs`.

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

File: api/server/httputils/write_log_stream.go [hunk 1]
```diff
@@ -85,6 +86,7 @@ func (s byKey) Swap(i, j int) {
 func stringAttrs(a backend.LogAttributes) string {
 	var ss byKey
 	for k, v := range a {
+		k, v := url.QueryEscape(k), url.QueryEscape(v)
```
Comment (aaronlehmann) [inline api/server/httputils/write_log_stream.go]: Why URL escaping? It looks like this is added to the log output stream, not a URL.

File: cli/command/service/logs.go [hunk 2]
```diff
@@ -84,6 +91,9 @@ func runLogs(dockerCli *command.DockerCli, opts *logsOptions) error {
 		maxLength    = 1
 		responseBody io.ReadCloser
 		tty          bool
+		// lf is used to delay the call to logs so that we can do some
+		// processing before we actually get the logs
+		lf func(context.Context, string, types.ContainerLogsOptions) (io.ReadCloser, error)
```
Comment (aaronlehmann) [inline cli/command/service/logs.go]: What does `lf` stand for?

Comment (dperny) [inline cli/command/service/logs.go]: log function

File: api/server/httputils/write_log_stream.go (hunk 1 above)
Comment (dperny) [inline api/server/httputils/write_log_stream.go]: URL escaping takes care of all of the special characters that might be present in the keys and values in a common and predictable way.

File: cli/command/service/logs.go [hunk 3]
```diff
@@ -212,43 +235,90 @@ type logWriter struct {
 	opts *logsOptions
 	f    *taskFormatter
 	w    io.Writer
+
+	// detailsCache saves a pre-sorted list of the details keys, so we don't
+	// have to recompute their order every time. this operates under the
+	// assumption that the details keys are always the same (they should be)
```
Comment (aaronlehmann) [inline cli/command/service/logs.go]: I'm not sure we can make this assumption.

File: cli/command/service/logs.go [hunk 4]
```diff
@@ -212,43 +235,90 @@ type logWriter struct {
...
+			// and then save the cache
+			lw.detailsCache = d
+		}
+		d := make([]string, 0, len(details))
+		for _, k := range lw.detailsCache {
+			d = append(d, fmt.Sprintf("%v=%v", k, details[k]))
 		}
+		// then join and append
+		output = append(output, []byte(strings.Join(d, ","))...)
+		output = append(output, []byte(" ")...)
```
Comment (aaronlehmann) [inline cli/command/service/logs.go]: `output = append(output, ' ')`

File: cli/command/service/logs.go [hunk 5]
```diff
@@ -212,43 +235,90 @@ type logWriter struct {
...
-		} else {
-			output = append(output, part...)
+			// then sort em
+			sort.Strings(d)
+			// and then save the cache
+			lw.detailsCache = d
+		}
+		d := make([]string, 0, len(details))
+		for _, k := range lw.detailsCache {
+			d = append(d, fmt.Sprintf("%v=%v", k, details[k]))
```
Comment (aaronlehmann) [inline cli/command/service/logs.go]: For type safety and efficiency, I prefer:

```go
d = append(d, k + "=" + details[k])
```

File: api/server/httputils/write_log_stream.go (hunk 1 above)
Comment (aaronlehmann) [inline api/server/httputils/write_log_stream.go]: How do we escape the other parts of the log message?

File: cli/command/service/logs.go [hunk 6]
```diff
@@ -212,43 +235,90 @@ type logWriter struct {
...
+	if lw.opts.timestamps {
+		output = append(output, parts[0]...)
+		output = append(output, []byte(" ")...)
+	}
+	// add the context, nice and formatted
+	formatted, err := lw.f.format(lw.ctx, logCtx)
+	if err != nil {
+		return 0, err
+	}
+	output = append(output, []byte(fmt.Sprintf("%s    | ", formatted))...)
```
Comment (aaronlehmann) [inline cli/command/service/logs.go]: The `Sprintf` is being used for simple string concatenation. It could be `formatted + "    | "`.

File: cli/command/service/logs.go [hunk 7]
```diff
@@ -212,43 +235,90 @@ type logWriter struct {
...
-		}
-
-		if i == contextIndex {
-			formatted, err := lw.f.format(lw.ctx, logCtx)
-			if err != nil {
-				return 0, err
+	// if we included timestamps, add them to the front
+	if lw.opts.timestamps {
+		output = append(output, parts[0]...)
+		output = append(output, []byte(" ")...)
```
Comment (aaronlehmann) [inline cli/command/service/logs.go]: output = append(output, ' ')

File: daemon/cluster/services.go [hunk 8]
```diff
@@ -430,9 +430,17 @@ func (c *Cluster) ServiceLogs(ctx context.Context, selector *backend.LogSelector
 				if err != nil {
 					m.Err = err
 				}
+				// copy over all of the details
+				for _, d := range msg.Attrs {
+					m.Attrs[d.Key] = d.Value
+				}
```
Comment (aaronlehmann) [inline daemon/cluster/services.go]: Rather than reformat these as a map, would it be possible to change the engine to use a slice internally? I think this would make things cleaner, since we end up sorting the attrs anyway, which involves going back to a slice. The round-trip through a map doesn't seem necessary. See #32418

File: cli/command/service/logs.go (hunk 3 above)
Comment (dperny) [inline cli/command/service/logs.go]: I actually think we can, for now? Only way I can see this not working is if the log stream contained many different services together, which would have different log options.

File: cli/command/service/logs.go (hunk 4 above)
Comment (dperny) [inline cli/command/service/logs.go]: Forgot about character syntax.

File: cli/command/service/logs.go (hunk 5 above)
Comment (dperny) [inline cli/command/service/logs.go]: Will do.

File: daemon/cluster/services.go (hunk 8 above)
Comment (dperny) [inline daemon/cluster/services.go]: I still don't disagree but I do think that change is out of scope for this PR.

File: api/server/httputils/write_log_stream.go (hunk 1 above)
Comment (dperny) [inline api/server/httputils/write_log_stream.go]: We don't need to. See #31956 for a more detailed (hehe) explanation.

File: cli/command/service/logs.go (hunk 3 above)
Comment (aaronlehmann) [inline cli/command/service/logs.go]: Perhaps in a future implementation, the set of keys could change over time. I think it's better not to encode this assumption into the client, especially if it's an optimization. I'd recommend either dropping the optimization, or finding a way to implement it without making the assumption (invalidate the cache if the set of keys changes, or something like that).

**Comment by dperny** [comment]
@aaronlehmann removed details sorting cache optimization, replaced it with comment describing how this could be optimized in the future.

File: client/parse_logs.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/pkg/errors"
+)
+
+// ParseLogDetails takes a details string of key value pairs in the form
+// "k=v,l=w", where the keys and values are url query escaped, and each pair
+// is separated by a comma, returns a map. returns an error if the details
+// string is not in a valid format
+// the exact form of details encoding is implemented in
+// api/server/httputils/write_log_stream.go
+func ParseLogDetails(details string) (map[string]string, error) {
```
Comment (aaronlehmann) [inline client/parse_logs.go]: I know you're probably tired of hearing me say this, but shouldn't this return a slice instead of a map? This is new code, and the caller immediately converts the map to a slice. The caller only sorts the keys right now, but if this returned something like `[]Attribute`, the caller could sort those (by key), and I think it would make the code simpler.

**Comment by aaronlehmann** [comment]
Design LGTM

File: client/parse_logs.go (hunk 9 above)
Comment (dperny) [inline client/parse_logs.go]: I actually use the map later on at `cli/command/service/logs.go L319` in `parseContext`.

Comment (dperny) [inline client/parse_logs.go]: I'm not tired of hearing it though. I'm definitely open to any time that we can change this from a map to a slice. In this particular instance, I do actually believe that a map is better than a slice because the purpose of `ParseLogDetails` is to turn log details into a machine-convenient format, not just as a gobetween for string conversions.

Comment (dperny) [inline client/parse_logs.go]: Like, this isn't just for internal use; I encourage the use of this as part of the SDK API and as the single source of truth for how to parse log detail k/v pairs.

Comment (aaronlehmann) [inline client/parse_logs.go]: Ah, okay

**Comment by dperny** [comment]
Experimental failure is most likely not my fault.

**Comment by aaronlehmann** [comment]
Needs a rebase.

**Comment by dperny** [comment]
@aaronlehmann rebased.

**Comment by dperny** [comment]
Stripped out CLI components and opened docker/cli#42.

**Comment by dperny** [comment]
i think i finally fixed the tests.

**Comment by aaronlehmann** [comment]
@dperny: I see a lot of vendored files removed. Can you confirm this is correct?

**Comment by dperny** [comment]
@aaronlehmann Rebased on master, which has those files removed now. DIff is more reasonable.

**Comment by aaronlehmann** [comment]
LGTM

File: integration-cli/docker_cli_service_logs_test.go [hunk 10]
```diff
@@ -277,7 +283,7 @@ func (s *DockerSwarmSuite) TestServiceLogsTTY(c *check.C) {
 	// and make sure we have all the log lines
 	waitAndAssert(c, defaultReconciliationTimeout, countLogLines(d, name), checker.Equals, 2)
 
-	cmd := d.Command("service", "logs", name)
+	cmd := d.Command("service", "logs", "--pretty=false", name)
```
Comment (thaJeztah) [inline integration-cli/docker_cli_service_logs_test.go]: Why do we have to add `--pretty=false` here? Did the default change?

Comment (dperny) [inline integration-cli/docker_cli_service_logs_test.go]: Yes. TTY logs require `--pretty=false` because we cannot reliably parse the details of TTY logs and therefore cannot pretty-print them. This is the only CLI usage change for the defaults.

Comment (thaJeztah) [inline integration-cli/docker_cli_service_logs_test.go]: Can we reverse that? Having a boolean flag that requires a value to be useful is really awkward. The output In 17.05 is the "pretty" version? Or different?

Alternative `--raw`?

Comment (dperny) [inline integration-cli/docker_cli_service_logs_test.go]: `--raw` might work. but you would still have to pass `--raw` with TTY logs. would that be acceptable?

**Comment by dperny** [comment]
currently trying to fix the circular dependency between the integration test in this commit and the corresponding pr in docker/cli.

**Comment by dperny** [comment]
i would just delete the damned integration test but i changed the cli for TTY containers and it'll break CI if we don't fix it.

**Comment by dperny** [comment]
Tests should go green now.

**Comment by aaronlehmann** [comment]
ping @fcrisciani: Wondering if the test failures here are related to #32906. I don't recall seeing that parciular failure before that PR (though maybe it was lost in the noise of the other failures we were experiencing). Now I've seen it in a few places. This test restarts the daemon several times.

File: client/parse_logs.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+	detailsMap := make(map[string]string)
```
Comment (cpuguy83) [inline client/parse_logs.go]: very minor but you can initialize this with `len(pairs)` lower down.

Comment (dperny) [inline client/parse_logs.go]: sure, yes.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by aaronlehmann** [comment]
From the log:

```
time="2017-05-12T02:12:47.514189530Z" level=error msg="Error getting nodes: rpc error: code = 2 desc = The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online." 
time="2017-05-12T02:12:47.514227394Z" level=error msg="Handler for GET /v1.30/nodes returned error: rpc error: code = 2 desc = The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online." 
```

About 30ms before this, a node is elected the leader. Might be a race involving the recent change to swarmkit that causes RPCs to be rejected if there is no leader. Perhaps it does not start accepting RPCs immediately after a successful leader election. I will look into it.

cc @cyli

**Comment by aaronlehmann** [comment]
Since the tests are being rerun, the test failure I'm talking about was:

```
02:12:52 ----------------------------------------------------------------------
02:12:52 FAIL: docker_cli_swarm_test.go:1371: DockerSwarmSuite.TestSwarmClusterRotateUnlockKey
02:12:52 
02:12:52 [d83e505dc0fca] waiting for daemon to start
02:12:52 [d83e505dc0fca] daemon started
02:12:52 
02:12:52 [d6be0e35b627d] waiting for daemon to start
02:12:52 [d6be0e35b627d] daemon started
02:12:52 
02:12:52 [d23c143787090] waiting for daemon to start
02:12:52 [d23c143787090] daemon started
02:12:52 
02:12:52 [d6be0e35b627d] exiting daemon
02:12:52 [d6be0e35b627d] waiting for daemon to start
02:12:52 [d6be0e35b627d] daemon started
02:12:52 
02:12:52 [d23c143787090] exiting daemon
02:12:52 [d23c143787090] waiting for daemon to start
02:12:52 [d23c143787090] daemon started
02:12:52 
02:12:52 [d6be0e35b627d] exiting daemon
02:12:52 [d6be0e35b627d] waiting for daemon to start
02:12:52 [d6be0e35b627d] daemon started
02:12:52 
02:12:52 [d23c143787090] exiting daemon
02:12:52 [d23c143787090] waiting for daemon to start
02:12:52 [d23c143787090] daemon started
02:12:52 
02:12:52 [d6be0e35b627d] exiting daemon
02:12:52 [d6be0e35b627d] waiting for daemon to start
02:12:52 [d6be0e35b627d] daemon started
02:12:52 
02:12:52 [d23c143787090] exiting daemon
02:12:52 [d23c143787090] waiting for daemon to start
02:12:52 [d23c143787090] daemon started
02:12:52 
02:12:52 docker_cli_swarm_test.go:1448:
02:12:52     c.Assert(err, checker.IsNil)
02:12:52 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc4205de220), Stderr:[]uint8(nil)} ("exit status 1")
02:12:52 
02:12:52 [d83e505dc0fca] exiting daemon
02:12:52 [d6be0e35b627d] exiting daemon
02:12:52 [d23c143787090] exiting daemon
02:12:55 
```

**Comment by fcrisciani** [comment]
@aaronlehmann I saw it once, but then checking it turns out that the swarm node ls fails because of missing quorum:
```
time="2017-05-12T02:12:47.513912921Z" level=debug msg="Calling GET /v1.30/nodes"                                                  
time="2017-05-12T02:12:47.514189530Z" level=error msg="Error getting nodes: rpc error: code = 2 desc = The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online."
time="2017-05-12T02:12:47.514227394Z" level=error msg="Handler for GET /v1.30/nodes returned error: rpc error: code = 2 desc = The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online."
time="2017-05-12T02:12:47.514797914Z" level=debug msg="Root CA set successfully" cluster.id=24lyiaypa4bfur1c6za2wct1o method="(*Server).UpdateRootCA" module=node node.id=1bsy8p9bht7rkdbh0e6x1i9f1
```
Not sure if it is a timing issue exposed by my patch or before the test was failing earlier with the crash. Ideas?

**Comment by aaronlehmann** [comment]
I think that in swarmkit, `ticksWithNoLeader` needs to be reset on a `Ready()` that provides a leader. I'll make this change in swarmkit.

**Comment by fcrisciani** [comment]
ok, lmk if there is something on my side needed

**Comment by thaJeztah** [comment]
Tests on this one are green now 👍
